sklearn/cluster/k_medoids_.py

```diff
@@ -0,0 +1,333 @@
+        Dataset to cluster.
+
+        Returns
+        -------
+        self
+        """
+
+        self._check_init_args()
+
+        # Check that the array is good and attempt to convert it to
+        # Numpy array if possible
+        X = self._check_input(X)
+
+        # Apply distance metric to get the distance matrix
+        distances = pairwise_distances(X, metric=self.distance_metric)
+
+        medoid_ics = self._get_initial_medoid_indices(distances,
+                                                      self.n_clusters)
+        cluster_ics = self._get_cluster_ics(distances, medoid_ics)
+
+        # Old medoids will be stored here for reference
+        old_medoid_ics = np.zeros((self.n_clusters,))
+
+        # Continue the algorithm as long as
+        # the medoids keep changing and the maximum number
+        # of iterations is not exceeded
+        self.n_iter_ = 0
+        while (not np.all(old_medoid_ics == medoid_ics) and
+                self.n_iter_ < self.max_iter):
+            self.n_iter_ += 1
+
+            # Keep a copy of the old medoid assignments
+            old_medoid_ics = np.copy(medoid_ics)
+
+            # Get cluster indices
+            cluster_ics = self._get_cluster_ics(distances, medoid_ics)
+
+            # Update medoids with the new cluster indices
+            self._update_medoid_ics_in_place(distances, cluster_ics,
+                                             medoid_ics)
+
+        # Expose labels_ which are the assignments of
+        # the training data to clusters
+        self.labels_ = cluster_ics
+
+        # Expose cluster centers, i.e. medoids
+        self.cluster_centers_ = X[medoid_ics]
+
+        # Expose intertia
+        self.inertia_ = self._compute_inertia(X)
+
+        # Return self to enable method chaining
+        return self
+
+    def _check_input(self, X):
+
+        X = check_array(X, accept_sparse=['csr', 'csc'])
+        # Check that the number of clusters is less than or equal to
+        # the number of samples
+        if self.n_clusters > X.shape[0]:
+            raise ValueError("The number of medoids %d must be larger "
+                             "than the number of samples %d."
+                             % (self.n_clusters, X.shape[0]))
+
+        return X
+
+    def _get_cluster_ics(self, distances, medoid_ics):
+        """Returns cluster indices for distances and current medoid indices"""
+
+        # Assign data points to clusters based on
+        # which cluster assignment yields
+        # the smallest distance
+        cluster_ics = np.argmin(distances[medoid_ics, :], axis=0)
+
+        return cluster_ics
+
+    def _update_medoid_ics_in_place(self, distances, cluster_ics, medoid_ics):
+        """In-place update of the medoid indices"""
+
+        # Update the medoids for each cluster
+        for cluster_idx in range(self.n_clusters):
+
+            if sum(cluster_ics == cluster_idx) == 0:
```

Comment:
There may also be eventual value in checking if a cluster did not change between iterations and hence avoiding some of this recalculation.